flocker/control/test/test_diffing.py

```diff
@@ -202,3 +203,140 @@ def test_different_uuids(self):
             wire_decode(wire_encode(diff)).apply(object_a),
             Equals(object_b)
         )
+
+
+class DiffTestObjInvariant(PClass):
+    """
+    Simple pyrsistent object with an invariant.
+    """
+    _perform_invariant_check = True
+    a = field()
+    b = field()
+
+    def __invariant__(self):
+        if self._perform_invariant_check and self.a == self.b:
+            return (False, "a must not equal b")
+        else:
+            return (True, "")
+
+
+class InvariantDiffTests(TestCase):
+    """
+    Tests for creating and applying diffs to objects with invariant checks.
+    """
+    def test_straight_swap(self):
+        """
+        A diff composed of two separate ``set`` operations can be applied to an
+        object without triggering an invariant exception.
+        """
+        o1 = DiffTestObjInvariant(
+            a=1,
+            b=2,
+        )
+        o2 = DiffTestObjInvariant(
+            a=2,
+            b=1,
+        )
+        diff = create_diff(o1, o2)
+        self.assertEqual(2, len(diff.changes))
+        self.assertEqual(
+            o2,
+            diff.apply(o1)
+        )
+
+    def test_deep_swap(self):
+        """
+        A diff composed of two separate ``set`` operations can be applied to a
+        nested object without triggering an invariant exception.
+        """
+        a = DiffTestObjInvariant(
+            a=1,
+            b=2,
+        )
+        b = DiffTestObjInvariant(
+            a=3,
+            b=4,
+        )
+        o1 = DiffTestObjInvariant(
+            a=a,
+            b=b,
+        )
+        o2 = o1.transform(
+            ['a'],
+            lambda o: o.evolver().set('a', 2).set('b', 1).persistent()
+        )
+        diff = create_diff(o1, o2)
+
+        self.assertEqual(
+            o2,
+            diff.apply(o1)
+        )
+
+    @capture_logging(assertHasMessage, DIFF_COMMIT_ERROR)
+    def test_error_logging(self, logger):
+        """
+        Failures while applying a diff emit a log message containing the full
+        diff.
+        """
+        o1 = DiffTestObjInvariant(
+            a=1,
+            b=2,
+        )
+        DiffTestObjInvariant._perform_invariant_check = False
+        o2 = o1.set('b', 1)
+        DiffTestObjInvariant._perform_invariant_check = True
```

Comment:
Ahh, I forgot about testtools.  Yea, maybe right.  In that case, at least try/finally to minimize the chance state gets left corrupted.